File: _11_ContainerWithMostWater.java
// space complexity - o(1)
    //time - o(n)
    //Ran on Leetcode successfully : Yes
    // Problem faced  : No
    //Approach : Use two pinter Approach from left and right, for current left and right water stored will be less of nums[left/rifgt]*right-left
    // whichever left or right is smaller , move it t the other direction until you come acroos the opposite side being smaller.
class Solution {
    public int maxArea(int[] height) {

        int res = Integer.MIN_VALUE;

        int start = 0 ;
        int end = height.length - 1;

        while(start < end){

            if(height[start] < height[end]){
                res = Math.max(res, height[start]*(end - start));
                start++;
            }
            else{
                res = Math.max(res, height[end]*(end - start));
                end--;
            }

        }
        return res;
    }
}

File: 11. Container with most water.py
#bruteforce solution
#Time Complexity : O(N)
#Space compleity : O(1)
class Solution:
    def maxArea(self, height: List[int]) -> int:
        left = 0
        right = len(height)-1
        print(left,right)
        max_volume = 0
        while(left<=right):
            h = min(height[left],height[right])
            max_volume = max(max_volume,h*(right-left))
            if height[left]<=height[right]:
                left+=1
            else:
                right-=1
        return max_volume